javascript/ql/src/semmle/javascript/TypeScript.qll

```diff
@@ -1474,6 +1474,41 @@ class ExternalModuleScope extends @externalmodulescope, Scope {
   override string toString() { result = "external module scope" }
 }
 
+/**
+ * TypeScript import declared in an explicit augmentation to global, 
+ * in the form of a variable of typeof the import.
+ * For example:
+ * import * as net_outer from "net"
+ * declare global {
+ * 		var net: typeof net_outer
+ * }
+ * 
+ * var s = net.createServer(); // this reference to net is an import
+ * 
+ * This import pattern was observed when analyzing the package web-component-tester
+ * https://www.npmjs.com/package/web-component-tester
+ */
+class TSGlobalDeclImport extends DataFlow::ModuleImportNode::Range {
+  string path;
+
+  TSGlobalDeclImport() {
+    exists(LocalVarTypeAccess q, ImportDeclaration i |
+      q.getContainer() instanceof GlobalAugmentationDeclaration and
+      i.(BulkImportDeclaration).getLocal().getVariable() = q.getVariable() and
+      path = i.getImportedPath().getValue() and
+      this = DataFlow::globalVarRef(q
+              .getEnclosingStmt()
+              .(VarDeclStmt)
+              .getADecl()
+              .getBindingPattern()
+              .getAVariable()
+              .getName())
+    )
```

Comment:
I think this can be written more neatly and more precisely as follows:

```
    exists(
      GlobalVariable gv, VariableDeclarator decl, TypeofTypeExpr tt, LocalVarTypeAccess pkg,
      BulkImportDeclaration i
    |
      // `gv` is declared with type `typeof pkg`
      decl.getBindingPattern() = gv.getADeclaration() and
      tt = decl.getTypeAnnotation() and
      pkg = tt.getExpressionName() and
      // `pkg` is imported as `import * as pkg from 'path'`
      i.getLocal().getVariable() = pkg.getVariable() and
      path = i.getImportedPath().getValue() and
      // this is a reference to `gv`
      this = DataFlow::globalVarRef(gv.getName())
    )
```

Reply:
Asger may have further suggestions for simplification or generalisation (e.g., are there any other kinds of imports we should consider?).

Reply:
Maybe we should avoid using `DataFlow::globalVarRef` here, as it creates a recursive dependency between `moduleImport` and `globalVarRef`, since `globalVarRef` already depends on `moduleImport` (but not the other way around (I think)).

The original version based on `GlobalAugmentationDeclaration` seemed fairly safe, but dealing with just global variables is bit more risky. I'll start an evaluation on typescript.slugs.

Reply:
That is a very good point. Even though performance looked fine in the evaluation, we should perhaps still go with `this = DataFlow::exprNode(gv.getAnAccess())` to avoid the cyclic dependency.